### PR TITLE
Add rst parse option for continuing after parse error

### LINKED ISSUE #8615: rstToHtml: Option to not throw error when parsing
Discussed in this issue from the nimforum: https://github.com/nim-lang/nimforum/issues/195

Basically, it would be awesome if rstToHtml could handle erroneous data without throwing an exception so that if the user makes an error in the RST, we can still show the generated RST up to a point in the code preview on the forum.

For example, if a stray backtick is entered, instead of erroring out, maybe the rstToHtml function could take in an `RstParseOption` like `NoErrors` or some other name that allows errors to happen but can still render. Maybe an easy solution would be to simply add a backtick at the end of the RST if one doesn't exist.

### PR DESCRIPTION
**Pull request by jyapayne**
Fixes: #8615

Adds the ability to specify a compiler flag to ignore rst parsing errors. This comes in handy when generating documentation or when the nim forum needs to render rst to html.

### CODE REVIEW

File: tests/nimdoc/invalidrst.nim [hunk 1]
```diff
@@ -0,0 +1,4 @@
+proc foo*() =
```
Comment (timotheecour) [inline tests/nimdoc/invalidrst.nim]: convention is to name helper modules not directly compiled with mXXX, eg `minvalidrst.nim`

File: lib/packages/docutils/rst.nim [hunk 2]
```diff
@@ -312,6 +312,14 @@ proc defaultMsgHandler*(filename: string, line, col: int, msgkind: MsgKind,
   if mc == mcError: raise newException(EParseError, message)
   else: writeLine(stdout, message)
 
+proc noExceptionMsgHandler*(filename: string, line, col: int, msgkind: MsgKind,
+                      arg: string) {.procvar.} =
+  let mc = msgkind.whichMsgClass
+  let a = messages[msgkind] % arg
+  let message = "$1($2, $3) $4: $5" % [filename, $line, $col, $mc, a]
```
Comment (timotheecour) [inline lib/packages/docutils/rst.nim:319]: import std/miscdollars
and use `toLocation` which I've added recently as the single place where location formatting is defined (to allow easy override)

Comment (jyapayne) [reply lib/packages/docutils/rst.nim:319]: Alright. Should I change the proc above as well?

**Comment by timotheecour** [comment]
I'm not yet convinced; can something like https://github.com/nim-lang/Nim/pull/14282 work instead? I can move `raiseParseError` to a reusable location; and add an optional file/line/col argument ; (this is meant to be usable in other contexts eg json parse errors)

do you have a minimal example where something like https://github.com/nim-lang/Nim/pull/14282 would not be sufficient?

> This comes in handy when generating documentation

IMO that should just error, but give a nice error with the sufficient context (possible with smthg like https://github.com/nim-lang/Nim/pull/14282 )

> or when the nim forum needs to render rst to html.

likewise

**Comment by jyapayne** [comment]
@timotheecour In nimterop, @genotrance and I would like the ability to include arbitrary C docs and have them render regardless of if they contain errors or not. #14282 is nice, but it would not allow for this feature since it would just error and not include the docs at all, which is annoying at best and sometimes a huge pain.

**Comment by timotheecour** [comment]
so I guess it relates to https://github.com/nim-lang/Nim/issues/13638 and https://github.com/nimterop/nimterop/issues/61.
Still not convinced this is the correct approach. How about this instead:
```
# add this:
--docfmt:fmt # fmt in rst,md,doxygen,raw,besteffort (for now just rst,raw)
# if not set, `--docfmt:rst` is implied (ie, set in newConfigRef())
```
right now only rst is supported but your use case would add `raw` (un-interpreted), which is easy to support and add.

That way, it allows you to check your assumptions about comment format, and if comments are in a different format to begin with, unless we have an easy way to convert to rst, there's no point in trying to format to rst in the 1st place (eg it could give wrong results yet make the parser pass without error)

also if you have concrete example file showing those C doc comments that would help this discussion

**Comment by jyapayne** [comment]
@timotheecour that would be ideal. I would love to just remove the rst formatting in the first place. I agree it works better and would be fine with that solution.

**Comment by timotheecour** [comment]
* so you could start with just `--docfmt:rst|raw` (should be fairly simple patch) and it'd solve at least this problem for now, with the advantage that it's extensible to other formats (eg, should someone start supporting `md` ; already have libs for that)

raw is un-interpreted so cannot raise

## note:
instead of a new cmdline flag, one option is a pushable pragma: so that nimterop can simply have this:
```nim
{.push docfmt:"raw".}
# .. nimterop generated code
{.pop.}
```
and users don't need to have a custom command line invocation to generate docs

**Comment by jyapayne** [comment]
> instead of a new cmdline flag, one option is a pushable pragma: so that nimterop can simply have this

I like that idea too, but have no idea how to implement it. I think for now it's simpler to do the `--docfmt` change.

**Comment by timotheecour** [comment]
the more I think of it the more the pragma makes sense. eg:
`nim doc --project main` would be incorrect with a global `--docfmt` if your project has modules that have classical doc comments (rst) and modules that have raw comments (eg nimterop wrapper).

This really is a module (or scope)-level annotation, not something a user should configure.

> I like that idea too, but have no idea how to implement it.

i can help if needed. As usual, easiest is to pick something that's similar in the compiler and track the code to see how it's implemented.

You need to propagate from a `{.docfmt:"raw".}` to a some flag that is accessible in `genComment(d: PDoc, n: PNode): string`.

One option is to restrict this feature to be at module scope for now, and then future implementations can relax that. So you could add a member in `TSymFlag` and attach it to the module where the pragma is defined.

Beyond the wrapper use case for foreign/C doc comments, this feature could in future be used to allow writing your doc comments in markdown in a given module (via `{.docfmt:md.}`) without breaking anything, which I'm sure many would appreciate.

In any case, check with @araq to make sure this can be accepted.

**Comment by Araq** [comment]
I don't like it, usually the documentation is already bad enough, adding an option for having even less quality is a bad idea. What does "Ignore parsing errors" even mean? That effectively adds an unspecified error recovery mechanism to the existing RST parser.

**Comment by jyapayne** [comment]
@Araq what about the option for `--docfmt=raw`?

**Comment by Araq** [comment]
Instead our RST/Markdown parser (yes, for the 100th time, it also does support markdown!) should get support for special syntactic construct where plaintext is inserted. Something like:

```.. code-block:: plaintext```

Or the same with triple backticks which I cannot write down here because markdown's triple backticks do not nest.

**Comment by kaushalmodi** [comment]
> because markdown's triple backticks do not nest.

They do.. increase the number of backticks in the nester block. e.g.

`````
This is wrapped in 5 back ticks
|   ````
|   This is wrapped in 4 back ticks
|   |   ```
|   |   This is wrapped in 3 back ticks
|   |   ```
|   ````
`````

**Comment by timotheecour** [comment]
> They do.. increase the number of backticks

TIL...


```nim
proc Mat*(rows: cint, cols: cint, `type`: cint) {.importcpp:"Mat::Mat(@)".} =
  ##[```
  @overload
  @param rows Number of rows in a 2D array.
  @param cols Number of columns in a 2D array.
  @param type Array type. Use CV_8UC1, ..., CV_64FC4 to create 1-4 channel matrices, or
  CV_8UC(n), ..., CV_64FC(n) to create multi-channel (up to CV_CN_MAX channels) matrices.
  ```]##
  runnableExamples: discard
```

image

not too bad?
so maybe what's needed is a way to convert a raw comment (eg from nimtero/treesitter) into something properly escaped so that it won't trip off rst (in case of odd nested triple backticks) but seems rare

@jyapayne WDYT ?

EDIT: it may also need to parse and remove the C++ comment markers (`/**` + `*` on each line), eg:
```
/**
 * A brief history of JavaDoc-style (C-style) comments.
 *
 * This is the typical JavaDoc-style C-style comment. It starts with two
 * asterisks.
 *
 * @param theory Even if there is only one possible unified theory. it is just a
 *               set of rules and equations.
 */
```
not sure whether this should be handled in stdlib, docgen, or some nimble package though

**Comment by jyapayne** [comment]
Hmm, I'll have to test it out!

File: lib/packages/docutils/rst.nim (hunk 2 above)
Comment (timotheecour) [reply lib/packages/docutils/rst.nim:319]: somehow I missed that; yes por favor! (or in a different "cleanup" PR depending on whether this is staling or not)

**Comment by jyapayne** [comment]
@timotheecour The triple backticks seems to work for now. Closing this

**Pull request closed without merging**